app/src/main/java/org/mozilla/fenix/components/FxaServer.kt

```diff
@@ -19,7 +22,27 @@ object FxaServer {
     fun config(context: Context): ServerConfig {
         val serverOverride = context.settings().overrideFxAServer
         val tokenServerOverride = context.settings().overrideSyncTokenServer.ifEmpty { null }
+
```

Comment:
There are basically five states here. I'll write out what servers we want to use for them:

1) regular builds - `Server.RELEASE`
2) regular builds where user set a `serverOverride` - `serverOverride`
3) mozilla online builds - `Server.CHINA`
4) mozilla online builds configured with a `serverOverride` - `serverOverride`.
5) mozilla online builds where user asked us to use a global server instead of "local"

In this PR, you're also looking at the old Fennec account state. If I understand correctly, I think what that's trying to achieve is support for state (4) for the migrated mozilla online users, e.g. users of mozilla online that had a custom server configured during a migration.

However, our FxA migration logic specifically detects custom servers and doesn't migrate them. So, these users (a vast, vast minority) won't have their FxA state carried over. They're expected to re-configure their custom servers and login again.

Cases (2) and (4) above are basically the same - we'll just use the override. So we just get:
1) any build configured with a `serverOverride` - use `serverOverride`
2) no `serverOverride`, regular builds - use `Server.RELEASE`
3) no `serverOverride`, mozilla online builds, domestic China servers allowed by user - use `Server.CHINA`
4) no `serverOverride`, mozilla online builds, domestic China servers not allowed by user - use `Server.RELEASE`

So, I think changes to `FxaServer.kt` in this PR can be reduced to the following snippet, which should cover everything:

```
    fun config(context: Context): ServerConfig {
        // If a server override is configured, use that. Otherwise:
        // - for all channels other than Mozilla Online, use Server.Release.
        // - for Mozilla Online channel, if domestic server is allowed, use Server.CHINA; otherwise, use Server.RELEASE
        val serverOverride = context.settings().overrideFxAServer
        val tokenServerOverride = context.settings().overrideSyncTokenServer.ifEmpty { null }
        if (serverOverride.isEmpty()) {
            val releaseServer = if (Config.channel.isMozillaOnline && context.settings().allowChinaDomesticServer) {
                Server.CHINA
            } else {
                Server.RELEASE
            }
            return ServerConfig(releaseServer, CLIENT_ID, REDIRECT_URL, tokenServerOverride)
        }
        return ServerConfig(serverOverride, CLIENT_ID, REDIRECT_URL, tokenServerOverride)
    }
```

Reply:
I need to explain a little bit here. 
For Fennec Mozilla Online build users, there are some of them using **Global** server. **Global** is not a default setting in either Fennec or Fenix for Mozilla Online builds, but it may not be thought as the custom server as well. 

So I read fxa.account.json here, trying to keep those Mozilla-Online-build && Global-server users' setting.

I have to say those Mozilla-Online-build && Global-server users should be the minority, but sorry for not having accurate statistics.

With this explanation, we can discuss more about this part. Both ways are acceptable for me.

Reply:
So, I wrote a response which github decided to erase... i'll rewrite it.

Thanks for clarifying!

A "custom server" is really just a ui concept. As far as the fxa module is concerned, there's just a `ServerConfig`. In a regular Fenix build, there is only one way to change this ServerConfig - via developer mode, which lets users input custom server URLs. In a MozillaOnline build, this PR also adds a flag to flip the Global server on. But the end result is the same as a user entering a custom server url - we end up modifying ServerConfig.

Regarding `fxa.account.json` - I really, really don't want to add custom migration logic into our main codebase. This is why we have the pretty extensive support/migration package in android-components - to keep everything organized there, so that we can easily delete all of this migration code as soon as we can. And at this point, I don't want to change that package, either. It's really just waiting for the long tail of users to finish migrating, then we delete it as soon as possible.

Most importantly, consider what already happened during the MozillaOnline migration to use Fenix codebase. Users that had Global server configured in Fennec all got logged-out: https://github.com/mozilla-mobile/android-components/blob/44013851e8e9b487e5a62ebd616775eed45d8a7c/components/support/migration/src/main/java/mozilla/components/support/migration/FennecFxaMigration.kt#L282

This PR adds a UI flag to switch-on Global servers, so I'm assuming that these users could only log-into regular Chinese servers. So let's assume they logged-into these Chinese servers. If you now migrate their Fennec preference, you will break these users - FxA module will try to talk to the Global server, while it's logged-into the Chinese server. It won't work, and they'll essentially get logged-out.

So, I don't think you can migrate this flag anymore. It should have been done during the initial migration, but now it's too late.

I would keep this PR simple - remove all of the `fxa.account.json` related code (that is, remove any custom migration logic).

Then, this PR will only do two things:
1) allow users to use Global servers via a UI flag in settings
2) pass along the `fxaExpectChinaServers` flag to migration logic.

See my other question about (2)... I'm assuming you must have had a custom fenix build that already set this flag, right?